File: packages/ios/Sources/ProtocolClient.swift
// SPEC: docs/protocol.md#PROTO-001/003..007 协议客户端（URLSession + 手写 SSE）
import Foundation

public final class ProtocolClient {
  public let base: String
  public var token: String?

  public init(base: String) {
    self.base = base.hasSuffix("/") ? String(base.dropLast()) : base
  }

  public enum ClientError: Error {
    case rpc(code: String, message: String)
    case http(Int)
    case malformed
  }

  private func call(method: String, payload: [String: Any]) throws -> [String: Any] {
    let env = Envelope.request(method: method, payload: payload)
    var req = URLRequest(url: URL(string: "\(base)/api/v1?method=\(method)")!)
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "content-type")
    if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "authorization") }
    req.httpBody = try JSONSerialization.data(withJSONObject: env)
    let sem = DispatchSemaphore(value: 0)
    var result: Result<[String: Any], Error> = .failure(ClientError.malformed)
    URLSession.shared.dataTask(with: req) { data, resp, err in
      defer { sem.signal() }
      if let err { result = .failure(err); return }
      guard let data else { result = .failure(ClientError.malformed); return }
      do { result = .success(try Envelope.parseResponse(data)) } catch { result = .failure(error) }
    }.resume()
    sem.wait()
    return try result.get()
  }

  // 认证（PROTO-003）
  public func handshake(deviceId: String, jwk: [String: String]) throws -> (nonce: String, caps: [String]) {
    let d = try call(method: "device.handshake", payload: ["deviceId": deviceId, "publicKeyJwk": jwk])
    return (d["nonce"] as? String ?? "", d["caps"] as? [String] ?? [])
  }

  public func bind(deviceId: String, nonce: String, password: String, signature: String) throws -> String {
    let d = try call(method: "device.bind", payload: ["deviceId": deviceId, "nonce": nonce, "password": password, "nonceSignature": signature])
    guard let t = d["deviceToken"] as? String else { throw ClientError.malformed }
    return t
  }

  public func bindTemporary(deviceId: String, password: String) throws -> String {
    let d = try call(method: "device.bindTemporary", payload: ["deviceId": deviceId, "password": password])
    guard let t = d["deviceToken"] as? String else { throw ClientError.malformed }
    return t
  }

  // 会话（PROTO-004）
  public func sessionList() throws -> [String: Any] { try call(method: "session.list", payload: [:]) }
  public func sessionHistory(_ id: String, max: Int = 50) throws -> [String: Any] {
    try call(method: "session.history", payload: ["sessionId": id, "maxMessages": max])
  }
  public func sessionCreate(workspaceId: String?) throws -> [String: Any] {
    try call(method: "session.create", payload: workspaceId.map { ["workspaceId": $0] } ?? [:])
  }
  public func prompt(_ id: String, text: String, visionNote: String? = nil) throws -> [String: Any] {
    try call(method: "session.prompt", payload: ["sessionId": id, "text": text, "visionNote": visionNote ?? NSNull()])
  }
  public func stop(_ id: String) throws -> [String: Any] { try call(method: "session.stop", payload: ["sessionId": id]) }
  public func models(_ id: String) throws -> [String: Any] { try call(method: "session.models", payload: ["sessionId": id]) }
  public func selectModel(_ id: String, provider: String, model: String, effort: String? = nil) throws -> [String: Any] {
    try call(method: "session.selectModel", payload: ["sessionId": id, "provider": provider, "model": model, "reasoningEffort": effort ?? NSNull()])
  }
  public func permissionGet(_ id: String) throws -> [String: Any] { try call(method: "permission.get", payload: ["sessionId": id]) }
  public func permissionSet(_ id: String, value: String) throws -> [String: Any] {
    try call(method: "permission.set", payload: ["sessionId": id, "value": value])
  }

  // 工作区/目录（PROTO-007）
  public func workspaceList() throws -> [String: Any] { try call(method: "workspace.list", payload: [:]) }
  public func listDirectory(_ path: String?) throws -> [String: Any] {
    try call(method: "host.listDirectory", payload: path.map { ["path": $0] } ?? [:])
  }
  public func createDirectory(path: String, name: String) throws -> [String: Any] {
    try call(method: "host.createDirectory", payload: ["path": path, "name": name])
  }
  public func workspaceCreate(path: String) throws -> [String: Any] {
    try call(method: "workspace.create", payload: ["path": path])
  }

  // 语音/视觉（PROTO-005/006）
  public func voiceTranscribe(audioBase64: String, format: String) throws -> String {
    let d = try call(method: "voice.transcribe", payload: ["audioBase64": audioBase64, "format": format])
    return d["text"] as? String ?? ""
  }
  public func visionDescribe(imageBase64: String, mime: String) throws -> String {
    let d = try call(method: "vision.describe", payload: ["imageBase64": imageBase64, "mime": mime])
    return d["text"] as? String ?? ""
  }

  /// SSE（PROTO-001）：逐行读取，data 帧交给回调；返回可取消的 Task
  public func events(onEvent: @escaping ([String: Any]) -> Void, onDisconnect: @escaping (String?) -> Void) -> Task<Void, Never> {
    Task.detached {
      var req = URLRequest(url: URL(string: "\(self.base)/api/v1/events")!)
      if let token = self.token { req.setValue("Bearer \(token)", forHTTPHeaderField: "authorization") }
      do {
        let (bytes, _) = try await URLSession.shared.bytes(for: req)
        var line = ""
        for try await byte in bytes {
          let c = Character(UnicodeScalar(byte))
          if c == "\n" {
            if line.hasPrefix("data: ") {
              if let data = String(line.dropFirst(6)).data(using: .utf8),
                 let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                onEvent(obj)
              }
            }
            line = ""
          } else { line.append(c) }
        }
        onDisconnect(nil)
      } catch {
        onDisconnect(error.localizedDescription)
      }
    }
  }
}

File: packages/android/app/src/main/java/dev/whalemaid/app/core/ProtocolClient.kt
// SPEC: docs/protocol.md#PROTO-001/003/004/005/006/007 协议客户端（OkHttp）
package dev.whalemaid.app.core

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.BufferedReader

class ProtocolClient(val base: String) {
  var token: String? = null
  private val client = OkHttpClient.Builder()
    .connectTimeout(5, java.util.concurrent.TimeUnit.SECONDS)
    .readTimeout(20, java.util.concurrent.TimeUnit.SECONDS)
    .build()
  private val json = Json { ignoreUnknownKeys = true }

  private val jsonMedia = "application/json; charset=utf-8".toMediaType()

  fun call(method: String, payload: Map<String, Any?>): JsonObject {
    val body = Envelope.request(method, payload).toString()
    val req = Request.Builder()
      .url("$base/api/v1?method=$method")
      .post(body.toRequestBody(jsonMedia))
      .apply { token?.let { header("authorization", "Bearer $it") } }
      .build()
    client.newCall(req).execute().use { res ->
      val text = res.body?.string() ?: ""
      return Envelope.parseResponse(text).getOrThrow()
    }
  }

  // 认证（PROTO-003）
  fun handshake(deviceId: String, publicKeyJwk: JsonObject): Pair<String, List<String>> {
    val d = call("device.handshake", mapOf("deviceId" to deviceId, "publicKeyJwk" to publicKeyJwk))
    return d["nonce"]!!.jsonPrimitive.content to (d["caps"]?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList())
  }

  fun bind(deviceId: String, nonce: String, password: String, nonceSignature: String): String =
    call("device.bind", mapOf("deviceId" to deviceId, "nonce" to nonce, "password" to password, "nonceSignature" to nonceSignature))["deviceToken"]!!.jsonPrimitive.content

  fun bindTemporary(deviceId: String, password: String): String =
    call("device.bindTemporary", mapOf("deviceId" to deviceId, "password" to password))["deviceToken"]!!.jsonPrimitive.content

  // 会话（PROTO-004）
  fun sessionList(): JsonObject = call("session.list", emptyMap())
  fun sessionHistory(sessionId: String, maxMessages: Int = 50): JsonObject =
    call("session.history", mapOf("sessionId" to sessionId, "maxMessages" to maxMessages))
  fun sessionCreate(workspaceId: String? = null): JsonObject =
    call("session.create", workspaceId?.let { mapOf("workspaceId" to it) } ?: emptyMap())
  fun prompt(sessionId: String, text: String, visionNote: String? = null): JsonObject =
    call("session.prompt", mapOf("sessionId" to sessionId, "text" to text, "visionNote" to visionNote))
  fun stop(sessionId: String): JsonObject = call("session.stop", mapOf("sessionId" to sessionId))
  fun models(sessionId: String): JsonObject = call("session.models", mapOf("sessionId" to sessionId))
  fun selectModel(sessionId: String, provider: String, model: String, reasoningEffort: String? = null): JsonObject =
    call("session.selectModel", mapOf("sessionId" to sessionId, "provider" to provider, "model" to model, "reasoningEffort" to reasoningEffort))
  fun permissionGet(sessionId: String): JsonObject = call("permission.get", mapOf("sessionId" to sessionId))
  fun permissionSet(sessionId: String, value: String): JsonObject = call("permission.set", mapOf("sessionId" to sessionId, "value" to value))

  /** 审批应答（REQ-008）：回显宿主稳定 rpcId */
  fun approvalRespond(rpcId: String, sessionId: String, approvalId: String, outcome: String): JsonObject =
    call("approval.respond", mapOf("rpcId" to rpcId, "sessionId" to sessionId, "approvalId" to approvalId, "outcome" to outcome))

  // 工作区/目录（PROTO-007）
  fun workspaceList(): JsonObject = call("workspace.list", emptyMap())
  fun listDirectory(path: String? = null): JsonObject = call("host.listDirectory", path?.let { mapOf("path" to it) } ?: emptyMap())
  fun createDirectory(path: String, name: String): JsonObject = call("host.createDirectory", mapOf("path" to path, "name" to name))
  fun workspaceCreate(path: String): JsonObject = call("workspace.create", mapOf("path" to path))

  // 语音/视觉（PROTO-005/006）
  fun voiceTranscribe(audioBase64: String, format: String): String =
    call("voice.transcribe", mapOf("audioBase64" to audioBase64, "format" to format))["text"]?.jsonPrimitive?.contentOrNull ?: ""
  fun visionDescribe(imageBase64: String, mime: String): String =
    call("vision.describe", mapOf("imageBase64" to imageBase64, "mime" to mime))["text"]?.jsonPrimitive?.contentOrNull ?: ""

  /** SSE（PROTO-001）：手动读取 event stream，data 帧交给回调 */
  fun events(onEvent: (JsonObject) -> Unit, onDisconnect: (String?) -> Unit): Thread {
    val thread = Thread {
      try {
        val req = Request.Builder()
          .url("$base/api/v1/events")
          .apply { token?.let { header("authorization", "Bearer $it") } }
          .build()
        client.newCall(req).execute().use { res ->
          val reader = BufferedReader(res.body!!.charStream())
          var line: String?
          while (reader.readLine().also { line = it } != null) {
            if (line!!.startsWith("data: ")) {
              try {
                onEvent(json.parseToJsonElement(line!!.substring(6)).jsonObject)
              } catch (_: Exception) { /* 坏帧忽略 */ }
            }
          }
        }
        onDisconnect(null)
      } catch (e: Exception) {
        onDisconnect(e.message)
      }
    }
    thread.isDaemon = true
    thread.start()
    return thread
  }
}
